### PR TITLE
Fix dock visibility issues

### LINKED ISSUE #66834: Editor dock rendering is broken
### Godot version

v4.0.beta2.official [f8745f2f7]

### System information

5.19.7-1-MANJARO with Plasma 5.25.5

### Issue description

Peek 2022-10-02 15-44
When I open a project, most of the editor's `docks` are not visible at first. They reveal only when I play around with the `Dock Positions`. Only then, their tab becomes visible (and I finally able to work with them).

And I am not sure if that's the same issue, but as you can see in the image, the (3-dot) handles for the `Dock Positions` are also sometimes in a "messy" spot (see the file dock on the lower left for example).

I didn't experience this issue with 3.x versions, but with every 4.0 version I've tried so far (alpha15, alpha16, beta1, beta2 (all from [https://downloads.tuxfamily.org/godotengine/4.0/](https://downloads.tuxfamily.org/godotengine/4.0/))

### Steps to reproduce

Start Godot 4.0 and open an existing or new project.

### Minimal reproduction project

_No response_

### LINKED ISSUE #84125: Godot crashes when changing Editor Features when certain docks are floating
### Godot version

master

### System information

Godot v4.2.beta (0ea75d77c) - Windows 10.0.22621 - Vulkan (Forward+) - dedicated NVIDIA GeForce RTX 2070 (NVIDIA; 31.0.15.3598) - Intel(R) Core(TM) i7-10750H CPU @ 2.60GHz (12 Threads)

### Issue description

When the import, filesystem, node, or history dock is floating, changing an editor feature crashes the engine.
Changing any setting triggers it.

This is because `EditorNode::_feature_profile_changed` assumes the docks are in a `TabContainer`, and if they are floating they aren't in one.

### Steps to reproduce

Make any of these docks floating: import, filesystem, node, or history
Open Editor Features (Editor -> Manage Editor Features...)
Make a profile and change a setting

### Minimal reproduction project

n/a

### PR DESCRIPTION
**Pull request by kitbdev**
Fix dock visibility for Editor Features and distraction free mode.

- fixes https://github.com/godotengine/godot/issues/66834
- fixes #84125 crash when changing Editor Features with floating docks (import, filesystem, node, or history)
- fixes error when closing floating dock if there are less tabs than before
- fixes loading layouts in distraction free mode being very buggy (bottom dock disappears, side docks might open and be empty)
- fixes floating dock still opened even if it was disabled in Editor Features when loading layout

Updated `TabContainer`'s `move_tab_from_tab_container` to preserve all tab data.

The crash fix is a workaround, ideally it wouldn't have to close all floating docks. I am working on a different pr that should fix this in the future.

There were 4 separate implementations to update the dock slots visibility:
`update_dock_slots_visibility()`,  `_update_dock_containers()`, in `_load_docks_from_config()`, and `_dock_tab_changed()` (removed the lines affecting bottom docks for this one, see #39177).
I have consolidated them all in `update_dock_slots_visibility()`.

### CODE REVIEW

File: scene/gui/tab_container.cpp [hunk 1]
```diff
@@ -381,17 +381,23 @@ void TabContainer::move_tab_from_tab_container(TabContainer *p_from, int p_from_
...
 	p_from->remove_child(moving_tabc);
 	add_child(moving_tabc, true);
 
 	set_tab_title(get_tab_count() - 1, tab_title);
 	set_tab_icon(get_tab_count() - 1, tab_icon);
+	set_tab_button_icon(get_tab_count() - 1, tab_button_icon);
 	set_tab_disabled(get_tab_count() - 1, tab_disabled);
+	set_tab_hidden(get_tab_count() - 1, tab_hidden);
 	set_tab_metadata(get_tab_count() - 1, tab_metadata);
+	get_tab_bar()->set_tab_icon_max_width(get_tab_count() - 1, tab_icon_max_width);
```
Comment (YeldhamDev) [inline scene/gui/tab_container.cpp:400]: Any reason why those are set before the child is moved to the right index?

Comment (kitbdev) [reply scene/gui/tab_container.cpp:400]: That's how it was done before, so I just kept it the same. I can change it to be after the move.

Comment (YeldhamDev) [reply scene/gui/tab_container.cpp:400]: Yes, as it would be preferable to use `p_to_index` instead of `get_tab_count() - 1` over and over.

**Comment by kitbdev** [comment]
Since this fixes a crash, should it be for 4.2?

**Comment by AThousandShips** [comment]
Can you please open an issue for the crash part, I don't see it reported in the linked issue, a crash is serious enough to track actively

**Comment by AThousandShips** [comment]
But note also that we're in the late stage of the 4.2 cycle and focus right now is on bugs introduced in 4.2 and fixing them so 4.2 introduces as few *new* bugs as possible, and as some of this behaviour goes back to 4.0 it's lower priority right now, marked for cherry picking for 4.2 so it can be added to it for 4.2.1 if it doesn't meet the deadline

**Comment by kitbdev** [comment]
Made an issue for the crash: https://github.com/godotengine/godot/issues/84125
If this isn't cherrypick-able I can create a smaller pr that is just the crash fix if needed.

**Review by YeldhamDev — APPROVED** [review]
Can't reproduce #66834 to confirm, but everything else seems alright.

**Comment by YuriSizov** [comment]
I'm curious if @kitbdev could reproduce https://github.com/godotengine/godot/issues/66834, or if this is an assumed fix based on code changes.

File: scene/gui/tab_container.cpp [hunk 2]
```diff
@@ -381,22 +381,29 @@ void TabContainer::move_tab_from_tab_container(TabContainer *p_from, int p_from_
...
 	move_child(moving_tabc, get_tab_control(p_to_index)->get_index(false));
+
+	set_tab_title(p_to_index, tab_title);
+	set_tab_icon(p_to_index, tab_icon);
+	set_tab_button_icon(p_to_index, tab_button_icon);
+	set_tab_disabled(p_to_index, tab_disabled);
+	set_tab_hidden(p_to_index, tab_hidden);
+	set_tab_metadata(p_to_index, tab_metadata);
+	get_tab_bar()->set_tab_icon_max_width(p_to_index, tab_icon_max_width);
+
```
Comment (YuriSizov) [inline scene/gui/tab_container.cpp:406]: This isn't really maintainable. I think we should create a method in `TabBar` that returns a copy of the entire `Tab` struct and add API to `TabContainer` which can fetch it and apply it to one of its tabs. It doesn't need to be exposed to scripting, just public for internal use.

File: editor/editor_node.cpp [hunk 3]
```diff
@@ -6593,6 +6513,10 @@ void EditorNode::_resource_loaded(Ref<Resource> p_resource, const String &p_path
 
 void EditorNode::_feature_profile_changed() {
 	Ref<EditorFeatureProfile> profile = feature_profile_manager->get_current_profile();
+	// Workaround: Close all floating docks to avoid crash.
```
Comment (YuriSizov) [inline editor/editor_node.cpp:6516]: Should be a FIXME comment.

File: editor/editor_node.cpp [hunk 4]
```diff
@@ -5078,36 +5054,41 @@ void EditorNode::_update_dock_slots_visibility(bool p_keep_selected_tabs) {
...
 		}
 
 		for (int i = 0; i < vsplits.size(); i++) {
-			bool in_use = dock_slot[i * 2 + 0]->get_tab_count() || dock_slot[i * 2 + 1]->get_tab_count();
+			bool in_use = dock_slot[i * 2 + 0]->is_visible() || dock_slot[i * 2 + 1]->is_visible();
 			if (in_use) {
 				vsplits[i]->show();
 			} else {
 				vsplits[i]->hide();
 			}
```
Comment (YuriSizov) [inline editor/editor_node.cpp:5089]: This can be simplified with `set_visible`.

File: scene/gui/tab_container.cpp (hunk 2 above)
Comment (YuriSizov) [reply scene/gui/tab_container.cpp:406]: To be clear, this is not a requirement for this PR, but with the number of things we need to keep track of, it's a bigger concern now.

**Review by YuriSizov — APPROVED** [review]
Seems like a good refactoring. I'm not sure about the workaround for the crash. There are no other reports of it, it seems, and this is not a proper fix. So maybe it's not urgent to fix in 4.2 (so the entire PR can be pushed to 4.3, with maybe a cherry-pick later).

**Comment by kitbdev** [comment]
I added some reproduction steps on https://github.com/godotengine/godot/issues/66834

I only found the crash since I was extracting all dock logic from `EditorNode` and found some issues and thought they should be applicable to 4.2, so I fixed them without the extraction.
I'm working on a different pr (open and close docks) that already has a proper fix for the crash.
So we can wait for that if we want to.

**Comment by YuriSizov** [comment]
Since this PR is pretty much ready (just too late for 4.2), I think that we should merge it first, and then discuss your other PR built on top of this.

**Comment by YuriSizov** [comment]
Cherry-picked for 4.2.2.

**Comment by YuriSizov** [comment]
Cherry-picked for 4.1.4.

**Pull request merged**